File: courses/codes/UnionFindExercise.java
import java.util.*;

class UnionFind {
    Map<String, String> R;

    public UnionFind() {
        this.R = new HashMap<>();
    }

    public void make(String x) {
        if (!R.containsKey(x)) {
            R.put(x, x);
        }
    }

    public String find(String x) {
        while (!x.equals(R.get(x))) {
            x = R.get(x);
        }
        return x;
    }

    public void union(String x, String y) {
        String rx = find(x);
        String ry = find(y);
        if (!rx.equals(ry)) {
            for (Map.Entry<String, String> entry : R.entrySet()) {
                if (entry.getValue().equals(ry)) {
                    entry.setValue(rx);
                }
            }
        }
    }
}

public class UnionFindExercise {

    public static void ex1() {
        Scanner scanner = new Scanner(System.in);
        List<String[]> friends = new ArrayList<>();

        while (true) {
            String line = scanner.nextLine();
            String[] split = line.split(" ");
            if (split.length == 1) {
                break;
            }
            friends.add(split);
        }

        String[] q = friends.remove(friends.size() - 1);

        UnionFind uf = new UnionFind();
        for (String[] pair : friends) {
            uf.make(pair[0]);
            uf.make(pair[1]);
            uf.union(pair[0], pair[1]);
        }

        System.out.println(uf.find(q[0]).equals(uf.find(q[1])) ? "yes" : "no");
    }

    public static boolean allValuesEqual(Map<String, String> R) {
        Set<String> values = new HashSet<>(R.values());
        return values.size() == 1;
    }
    
    public static boolean containsAllFrom1ToN(Map<String, String> m, int N) {
        for (int i = 1; i <= N; i++) {
        	String s = String.valueOf(i);
        	if (!m.containsKey(s))
        		return false;
        }
        return true;
    }
    
    

   

    public static void main(String[] args) {
        ex1();
    }
}

File: courses/codes/union_find_exercise.py
import copy

# simplest UnionFind implementation with Arrays/Lists
class UnionFind:

    def __init__(self):
        self.R = dict()

    def make(self, x):
        if x not in self.R:
            self.R[x] = x

    def find(self, x):
        return self.R[x]

    def union(self, x, y):
        rx = self.find(x)
        ry = self.find(y)
        if rx != ry:
            R_copy = copy.deepcopy(self.R)
            for k, v in R_copy.items():
                if v == ry:
                    self.R[k] = rx


def ex1():

    line = input().split(" ")
    friends = []
    while len(line) > 1:
        friends.append((line[0], line[1]))
        line = input().split(" ")
    q0, q1 = friends.pop(-1)

    uf = UnionFind()
    for f0, f1 in friends:
        uf.make(f0)
        uf.make(f1)
        uf.union(f0, f1)

    print('yes') if uf.find(q0) == uf.find(q1) else print('no')

# return true if all values of the dictionary d are equal; false o.w.
def all_values_equal(d):
    values = list(d.values())
    val = values[0]
    for i in range(1, len(values)):
        if val != values[i]:
            return False
    return True




ex1()
